### DISCUSSION TITLE
Are there any good dependency injection libraries available/compatible with bun?

### DISCUSSION THREAD

**Discussion by jdgriffith** [oven-sh/bun#3273] [Q&A]
I'm coming from nestjs and am comfortable with their dependency injection (DI) or inversion of control, which works incredibly well. I have decent sized codebases with different modules and would like to keep some of the same structure if I port the code to the bun runtime. Are there any options for someone like me?

P.S. I don't need http capability atm, only need DI.

**Comment by iamyuu** [comment]
I try using [typedi](https://github.com/typestack/typedi) and it work, but you need to use `Container.get`

<details>
<summary>example code</summary>

```ts
import 'reflect-metadata';
import { Service, Container } from 'typedi';

@Service()
class AppRepository {
	foo = () => {
		console.log('bar');
	};
}

@Service()
class AppService {
	baz = () => {
		Container.get(AppRepository).foo();
	};
}

new AppService().baz();
```

</details>

**Reply by algoflows** [reply]
Any more examples?

**Comment by struckchure** [comment]
Currently trying this in Elysia and it doesn't work, but using it as stated in that example w/o elysia, works fine

Ideas?

**Reply by aquapi** [reply]
Why do you need these patterns in Elysia?
It's not designed to be used like that.

**Reply by struckchure** [reply]
I want to use di for my app services outside elysia, so I can do proper unit testing

or is there a way to inject dependencies like db connections to respository/services without props drilling from the controller (elysia)?

**Reply by aquapi** [reply]
You know many can do proper testing without using OOP patterns right?
You can inject using something like `.decorate` I think

**Reply by struckchure** [reply]
You're right, but i'm used to having DI, coming from NestJs.

**Reply by nad-au** [reply]
I spent nearly two years building APIs with Nest, and I was happy with result coming from OOP with .NET. However, since using Bun & Elysia, I'm beginning to accept more functional-style programming. And indeed with plugins and decorate, you can achieve similar to DI.
